### ISSUE TITLE
Fix minetest.registered_nodes so it will be reliable

### ISSUE THREAD

**Issue by dsohler** [minetest/minetest#5267]
I am currently experimenting with a mod that takes `minetest.registered_nodes` and iterates over it with a for loop and creates nodes basing on some filters and defined options.Every now and then I get unknown nodes placed in the world where nodes are placed that were generated before by my mod based on what’s in `minetest.registered_nodes`.

I can reproduce it using [Craftguide](https://github.com/minetest-mods/craftguide) in combination with [my mod](https://github.com/dsohler/xtend_modpack/tree/master/xfurniture).

One of the registered nodes (see name in debug information)

![screenshot_20170218_145202](https://cloud.githubusercontent.com/assets/1668792/23093661/b6348cf2-f5e9-11e6-8b35-b0ac6c212067.png)

After enabling Craftguide the node becomes unknown (see error message after punching: same name)

![screenshot_20170218_145236](https://cloud.githubusercontent.com/assets/1668792/23093660/b63414d4-f5e9-11e6-9cfc-0452896488b2.png)

It happens with other mods, too. It happens without any other mods. It randomly happens every now and then. I can reproduce it 100% with the Craftguide mod. (I assume because of this mod also iterates over all nodes and causes delay, or something.)

**Comment by rubenwardy** [comment]
If you type /mods, is xfurniture there?

This looks more like a mod bug, to me

**Comment by dsohler** [comment]
It is. As seen in the screenshot some of the other nodes created by the mod are there.

The loop looks like this.

    for name,def in pairs(minetest.registered_nodes) do
        if (def.drawtype == 'normal') then
            -- register nodes here.
        end
    end

Sometimes it does not register the nodes. And under other circumstances it does.

Possible race condition?

**Comment by HybridDog** [comment]
minetest.registered_nodes has a special metatable.
Actually (almost) no node has drawtype = "normal" in it's node definition.
Use rawget(def, "drawtype") if you want to get the value which is passed when registering the node.
Maybe the Craftguide somehow removes the metatable from minetest.registered_nodes, which means a bug in the Craftguide mod.

**Comment by sofar** [comment]
I think you have a timing issue. You're listing all existing mods, but your code runs before some other mods run. The loading order of mods is random unless you add depends.txt entries.

You will need to explicitly depend on any mod that you want to do something with.

**Comment by kilbith** [comment]
> Maybe the Craftguide somehow removes the metatable from minetest.registered_nodes, which means a bug in the Craftguide mod.

The only way to remove a metatable in Lua is to empty its attached table, like so : `foo = {}`. Which `craftguide` does not do.

**Comment by dsohler** [comment]
> You will need to explicitly depend on any mod that you want to do something with.

This is exactly what I do.

I depend my mod on `default` and sometimes nodes starting with `default:` are not present when loading and thus previously placed nodes basing on said `default:` nodes become invalid.

**Comment by sofar** [comment]
other mods may add nodes that start with `:default:` .

**Comment by dsohler** [comment]
Here is another example:

With Craftguide enabled (i use Craftguide because I guess it creates the most load when loading because it iterates, too.) The nodes missing are `default:stone_with_iron` and `default:bronzeblock`

![screenshot_20170219_012831](https://cloud.githubusercontent.com/assets/1668792/23097958/8737267c-f642-11e6-835f-865421efc8d6.png)

Craftguide disabled:

![screenshot_20170219_012853](https://cloud.githubusercontent.com/assets/1668792/23097959/87376a56-f642-11e6-8154-fa3f15f505ef.png)

Are those two not in the mod called `default` when refering to it in the `depends.txt` file?

**Comment by sofar** [comment]
where is your code exactly?

**Comment by dsohler** [comment]
https://github.com/dsohler/xtend_modpack/tree/master/xfurniture

**Comment by sofar** [comment]
Hmm, it appears indeed that the registration array isn't reliable while in mod init phase.

I could also see that mods performing overrides pose a problem, since they potentially remove and insert modified items?

**Comment by dsohler** [comment]
> Hmm, it appears indeed that the registration array isn't reliable while in mod init phase.

So it's a bug, and not "invalid, not a bug"

**Comment by numberZero** [comment]
@dsohler You can’t iterate over a table you’re modifying! That’s documented! http://www.lua.org/manual/5.1/manual.html#pdf-next

**Comment by red-001** [comment]
the loading of mods is sequential  and there is only one lua thread so race conditions are impossible. It sounds like there is an error in your mod.

**Comment by numberZero** [comment]
@red-001 `register_node` changes `registered_nodes`, so it’s an error to call it from within a loop iterating over `registered_nodes`.

**Comment by dsohler** [comment]
My mod does not use `next()`. It uses a for loop.

Here’s a quick example that it doesn’t really matter when iterating over a table that is altered during the iteration. The for loop ALWAYS iterates over all entries that are initially in the table.

[Rename this file to test.lua](https://github.com/minetest/minetest/files/786172/test.txt) and run it.

~~~~
$ lua5.2 test.lua 
The table initially has 3 entries.

Iteration 1: table[1] = foo
Iteration 2: table[2] = bar
Iteration 3: table[3] = baz
Iterated over all initial entries.

Iteration 4: table[7] = added during iteration 2
Iteration 5: table[12] = added during iteration 1
Iteration 6: table[11] = added during iteration 4

The table now has 8 entries
table[1] = foo
table[2] = bar
table[3] = baz
table[5] = added during iteration 6
table[7] = added during iteration 2
table[13] = added during iteration 5
table[12] = added during iteration 1
table[11] = added during iteration 4
~~~~

That proven: If the loop does not iterate over an entry in the table then it is because the entry does not exist in the table in the first place.

**Comment by rubenwardy** [comment]
pairs calls next

**Comment by numberZero** [comment]
yeah, sure, whatever...
1. `for` loop uses the iterator you pass to it, that is, what is returned by the `pairs` function, that is, the `next` function.
2. Such error may or may not happen, that depends solely on Lua internals. It probably rehashes the table when it grows to a certain size, thus breaking consequent iteration.

@rubenwardy returns, not calls

**Comment by red-001** [comment]
Yeah you should always avoid inserting or removing entries in a table while iterating over it. iirc there are ways around the issue but you are better off just avoiding it altogether.

**Comment by HybridDog** [comment]
>It probably rehashes the table when it grows to a certain size, thus breaking consequent iteration.

So actually it indirectly is a race condition, isn't it?

Someone had another loop problem, this behaviour transpired for him/her but not for me:
```
for i,v in pairs(some_table) do
	some_table[i] = nil -- v becomes nil here, too
end
```

**Comment by dsohler** [comment]
@HybridDog It continues adding and iterating until at some point pairs becomes faster than the adding of new entries to the table.

[See my script here](https://github.com/minetest/minetest/issues/5267#issuecomment-280946924)

**Comment by numberZero** [comment]
@HybridDog 
> v becomes nil here, too

it doesn’t

```lua
v = "xxx"
some_table = {a=1, b="ddd", c=33}
for i,v in pairs(some_table) do
  some_table[i] = nil
  print(v) -- prints the value
end
print(v) -- prints "xxx"
```

> race condition

Not exactly. That wasn’t the time what were determining the result, as everything were happening in a single thread, but only the initial sate (at the beginning of the `init.lua`). Such things are way easier to debug (as they don’t disappear on step-by-step execution), although may be similarly hard to reproduce (as the initial state may vary seemingly randomly).

**Comment by rubenwardy** [comment]
This two `v`s are different variables - different scope

**Comment by numberZero** [comment]
@rubenwardy sure. But neither is `nil`.

**Comment by HybridDog** [comment]
> it would look like v became nil due to the assignment

The assignment usually doesn't set v to nil but it happened to Ombridide. It makes code behave inconsistently, maybe it's an optimization, which gives v and some_table[i] the same lua variable instead of copying it.

>as everything were happening in a single thread

The garbage collector runs in a separate thread (afaik), l assume it influences the decision about when the table becomes rehashed, which indirectly makes it a race condition.
